### ISSUE TITLE
Meta sends escape enabled while running filter from "autocmd"

### ISSUE THREAD

**Issue by antofthy** [vim/vim#5617]
**Describe the bug**
After upgrading to  Vim-8.2 the 'xterm modifyOtherKeys' patch has caused issues with me...  I use "autocmd" to read in a binary encrypted file, which is then decrypted using a password read directly from the users TTY.

If I read from the xterm TTY from the command run from "autocmd" and I type 'A' I get the input '\e[27;2;65~'  which is the ANSI code for a 'Shifted-A' and not an actual 'A'.   As you can imagine the password is then wrong, the the command fails to decrypt the file I want to edit.  Only a problem if the passwords contain shifted characters in it!

If I use a :%!command directly, there is no problem...  So VIM is disabling the 'xterm modifyOtherKeys' it for that.   It only becomes a problem when I do this from a "autocmd".

This sort of "autocmd" input is what would be used by commands like "gpg" and "openssl enc" to encrypt and decrypt files via vim "autocmd"  and this is currently failing!

**To Reproduce**
run `vim file.test` in a xterm does not matter if the file does not exist

Inside vim...
```vim
:" check t_TI is set to t_TI=^[[>4;2m
:set t_TI
:" Set an autocommand to read from TTY
:autocmd  BufWritePre,FileWritePre   *.test !bash -c 'cat </dev/tty >/dev/tty'
:" now write the file to run the above autocmd
:w
```
The command will not be waiting from some input from the TTY...
type 'A'  and you will see the resulting ANSI sequence.

You can not even type ctrl-D to terminate the above, and will need to kill the "cat" from another terminal!

If you did the command directly without using autocmd,  EG:
```vim
" Direct shell call
:!bash -c 'cat </dev/tty >/dev/tty'
```
It works perfectly fine.  The problem only happens from autocmd, for both call '!' or filter '|' shell commands.

**Expected behavior**
Input should be what the user typed, not ANSI escape sequences.

**Environment (please complete the following information):**
Fedora Packages:
vim-common-8.2.109-1.fc31
vim-enhanced-8.2.109-1.fc31
vim-filesystem-8.2.109-1.fc31

vim --version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan 10 2020 00:00:00)
Included patches: 1-109
Modified by <bugzilla@redhat.com>
Compiled by <bugzilla@redhat.com>

Terminal   xterm-351-1.fc31
TERM=xterm-256color

**Additional context**
This appears to be the reason the patch was added...
https://github.com/vim/vim/issues/4974#issuecomment-541431801

A temporary fix was provided in
https://github.com/vim/vim/issues/5200  and https://github.com/vim/vim/issues/5612#issuecomment-584747235

**Comment by brammool** [comment]
it is now possible to temporarily disable modifyOtherKeys with: call echoraw(&t_TE)
And enable it again with:  call echoraw(&t_TI)

**Comment by antofthy** [comment]
@brammool Added to the functions my autocmd's call for encrypted files...
   "call echoraw(&t_TE)

E117: Unknown function: echoraw

At the moment the only solution I have is to disable it entirely, or output the ANSI sequence myself in a password reader wrapper (non-vim solution).

ASIDE...
Vim Encrypted file handler...
   https://antofthy.gitlab.io/software/#encrypt.vim

**Comment by k-takata** [comment]
Do you use 8.2.0258 or later?

**Comment by antofthy** [comment]
Update from fedora repo is only until 8.2.236
and that is only on my local machine.  Not on others.

But I already have a temporary (non-vim) solution, Just looking to get this fixed for the future without needing any extras.  EG: the filters called by the autocmd's do the right thing.  That way after a period of time, after all machines have updated I can get rid of my temporary solution, and forget about this 'weird period'.

**Comment by antofthy** [comment]
Something is very strange about these ANSI control sequences as defined!
echoing the "t_TE" control sequence before getting user input was causing the first letter
to the prompt that follows to 'disappear'!

So I did some tests of the ANSI sequences vim uses...
The 'echo' command are being performed just on the command line bash, not from in vim itself.

For example this loses the first letter 'A':     `echo $'\e[>4mABCD'`
And this prints none of the numbers!:      `echo $'\e[>4m1234'`
Adding a RETURN does not help...       `echo $'\e[>4m\r1234'`
Nor does adding a NEWLINE... `echo $'\e[>4m\n1234'`
The newline IS output, but NOT the numbers!

As it is missing one letter adding a 'junk' character to the ANSI control sequences seems to fix the problem...    For example:     `echo $'\e[>4m~1234'`     or    `echo $'\e[>4m~ABCD'`

I think I may be missing something about ANSI sequences!  or XTerm implements it wrong!
But if I am right, then the vim termcap settings are not correct.

So is this a "XTerm" fault, or a "vim" fault (mis-read of ANSI sequences)?

**Comment by chrisbra** [comment]
I suppose you were using `:!echo ...`? I cannot reproduce this.

**Comment by antofthy** [comment]
This is was in normal bash shell (editing previous comment to that effect), just testing the control sequences VIM is using for t_TI and t_TE termcap entries.  Run them on the command line to see the effects of the ANSI escape sequence.

Basically either XTerm is not terminating the sequence on the 'm', or VIM needs to output an extra 'junk' character.

**Comment by antofthy** [comment]
**Update...**

The extra character required weirdness in the implementation of the ANSI sequences in XTerm, as shown above appears to be fixed now (Fedora 31).   That seems to be a non-vim problem.

But the original problem, as demonstrated in the original first post, where the 'xterm modifyOtherKeys' is still enabled while running an autocmd filter that reads from a terminal, **remains**!

**Current solutions...**
* And I can now use echoraw() to wrapper the filter function call within the autocommand.
as used in autocmd vim function in https://antofthy.gitlab.io/software/#encrypt.vim
```
call echoraw(&t_TE)
"... do the terminal input filter
call echoraw(&t_TI)
```

* I can also disable the use of modifyOtherKeys completely using
```
if (v:version >= 820)
  let &t_TI = "<Csi>>2m"
  let &t_TE = ""
endif
```

* Or ensure input program disables the 'xterm modifyOtherKeys' mode.
For example from a bash script using something like...
``` bash
if [[ $RTYPE =~ FC3[12] ]] && tty >/dev/null; then
  #echo >/dev/tty "Fixing XTerm ANSI 'ModifyOtherKeys'..."
  case "$TERM" in
  xterm*) echo -n $'\e[>4n' >/dev/tty ;;
  esac
fi
```

In fact I am actually doing all three, to ensure this debug does not interfere with the terminal input (reading a password, with echoed stars to decrypt encrypted files for editing).

**Comment by antofthy** [comment]
**Please note however that I disagree with the closure of this bug.**

Adding the ability to use echoraw() does NOT solve the bug, it only provides a work around for the bug!

Current version of vim...  8.2.694-1.fc31

**Comment by brammool** [comment]
OK, I found the spot where the behavior is different when autocommands are executed.  Hmm, we can't just change that without lots of side effects.  One solution is to do what we do for bracketed paste.  That seems like an acceptable solution.

**Comment by Shane-XB-Qian** [comment]
@brammool though not very sure about the issue desc here, but probably better revert the patch#8.2.0767, looks it made vim (ater this patch) a little mess ..
it probably fixed this issue in pure xterm terminal, but made a little mess in e.g xfce-term terminal ..
e.g: vim --clean , then just ':' , it output the 'seq' sth ..., and output a lot in somewhere if you had a little complicated vimrc .. 
wish you can revert ... revert ... -_-# or looks lots compatible work need to do to fit this change ...

**Comment by Shane-XB-Qian** [comment]
btw: indent in this term.c looks is broken, mixed to use space and tab ..

**Comment by brammool** [comment]
This should be fixed by 8.2.0767.
I do not understand the problem from Shane-XB-Qian .
I don't see a problem with indenting in term.c

**Comment by Shane-XB-Qian** [comment]
lots output if startup with a little complicated vimrc, or e.g:
1, vim --clean
2, just ':'
![image](https://user-images.githubusercontent.com/13550529/82128257-55749080-97ec-11ea-8fde-eaa470bc61b6.png)

**Comment by Shane-XB-Qian** [comment]
pls check the indent if opened with a different editor (other than vim) probably would be more clear to see the broken indent ... :)
![image](https://user-images.githubusercontent.com/13550529/82128316-cae06100-97ec-11ea-88da-09b1e5d756c8.png)

**Comment by Shane-XB-Qian** [comment]
pls check the bottom of those 2 screen-cut, the mess of output (example) is like that ... :)
but: you would not see those if opened with a 'pure' xterm terminal ..
1. xterm
2. vim --clean
3. just ':'

**Comment by brammool** [comment]
Still don't see any problem with the indent.
Can't reproduce the output when pressing ":"...

**Comment by Shane-XB-Qian** [comment]
you are using a pure 'xterm' ? try other emulator like xfce-term, or sth else ..
what i remembered, (before this patch) modifyotherkey is default on when pure xterm, but not xfce-term ..

as for indent, for example: 
line#3498 is 4 spaces as its indent, but others were a-8-spaces-len-tab ..
line#3472 is a tab plus 4 spaces, line#3473 is 2 tabs ..

**Comment by brammool** [comment]
Still cannot reproduce the ">4;2m" codes.  What is your t_TI set to?

Indent looks fine: line 3472 has 12 spaces worth, next line 16, which is correct.

**Comment by Shane-XB-Qian** [comment]
i did not set t_TI but the default value is `^[[>4;2m`

for indent: yes, the len is correct, but probably should not mixed to use them, it would be disaster e.g if temp open it with a pure text editor .. or let it be, a tip but up to you, just a fyi.

whatever, but the 'seq' output is an issue, wish you can revert (or refine), or was there any tricky here ? or i had to find a way to compatible this - this would be terrible .. :(

**Comment by Shane-XB-Qian** [comment]
i saw this had been refined with a new patch, thanks @brammool , that ':' output issue fixed.

but another one - mess output when vim 'startup' with a little complicated vimrc - still there .. 
as a workaround, for now to me / i think just to disable modifyotherkey: `let &t_TI = ""` and `let &t_TE = ""`, or if someone needed it ..

btw: to check modifyotherkey whether enabled/disabled by `CTRL-SHIFT-V CTRL-V` in vim-doc looks was not accurated, since in some os: 'CTRL-SHIFT-V' is a shortcut to paste when insert mode, so my comment said modifyotherkey on xfce-term perhaps was not correct, it had no chance to see code after `CTRL-SHIFT-V CTRL-V` unless disabled it ..

**Comment by brammool** [comment]
That basic problem is still that t_TI and t_TE are set with sequences that the terminal doesn't understand.  We can add some workarounds, but we can't fix that problem.  It most likely is caused by setting $TERM to a value starting with "xterm" while the terminal is not actually an xterm.

**Comment by Shane-XB-Qian** [comment]
yea, i see ..
but here, now (after patch#767) even prepended by `:silent` it would output such 'seq' still (if did not reset t_ti / t_te) .. that's so annoying .. for now, if someone like modifyotherkey on, she/he had to take such mess output (/annoying or c-l).. probably not a good situation ..

**Comment by antofthy** [comment]
@brammool 
>OK, I found the spot where the behavior is different when autocommands are executed. Hmm, we can't just change that without lots of side effects. One solution is to do what we do for bracketed paste. That seems like an acceptable solution.

Can you elaborate...  Was this done?  Does it mean you apply t_te off before a shell call (filter or otherwise) from autocmd, and turn redo t_ti again afterwards?

That to me would be the logical solution as you are effective leaving 'vim' and giving the shell command a 'sane' terminal for the duration.  It looks as if that is what vim is doing for a normal direct (non-autocmd) shell call (command or filter), just not from autocmd.

**Comment by brammool** [comment]
Anthony Thyssen wrote:

> @brammool > OK, I found the spot where the behavior is different when
> autocommands are executed. Hmm, we can't just change that without lots
> of side effects. One solution is to do what we do for bracketed paste.
> That seems like an acceptable solution.
> 
> Can you elaborate...  Was this done?  Does it mean you apply t_te off
> before a shell call (filter or otherwise) from autocmd, and turn redo
> t_ti again afterwards?
> 
> That to me would be the logical solution as you are effective leaving
> 'vim' and giving the shell command a 'sane' terminal for the duration.
> It looks as if that is what vim is doing for a normal direct
> (non-autocmd) shell call (command or filter), just not from autocmd.

Yes, this was implemented.  And then side effects had to be fixed.

There are several ways to temporarily leave Vim.  Some are slower, thus
are avoided when the shell command is not supposed to have user
interaction.  Using the pty directly to prompt for a password is an edge
case, I hope it works now.  Still, it can result in the terminal to be
changed (e.g. scrolled up) without Vim knowing about it, and since
redrawing the whole screen is slow this is sometimes omitted and you end
up with a messed up display.  It's a compromise.

-- 
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?
A: Top-posting.
Q: What is the most annoying thing on usenet and in e-mail?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Shane-XB-Qian** [comment]
@brammool  is it possible to check that 'seq' was _ needed/recognized or `out_str(T_CTI)` did (or not) _ before `out_str(T_CTE)` ? if yes, such side-effect maybe able to pass .. or had to do that i said workaround which just set modifyotherkey (t_te/ti) off if did not like such annoying output and not like c-l ..
just a thinking .. and if needed, i can open a new separate ticket - if interrupt this @antofthy original tkt here too much .. :)
or let it be, so far, issue still there, but can be workaround ..

**Comment by Shane-XB-Qian** [comment]
@brammool  looks mostly cases had been overcome by you, but one left so far i noticed: 
`let sqx = system('ls')` such shell call still output seq .. perhaps should overcome it as well, then no need some ugly workaround .. fyi.

**Comment by antofthy** [comment]
The changes have come though, via the Fedora Repos, and all is good... Thank you @brammool

**Comment by Shane-XB-Qian** [comment]
not really, e.g echo system('ls') or assign to a var, it still output such seq .. fyi.
though it perhaps not related your original tkt, but started from from such fix.. looks ..

**Comment by chrisbra** [comment]
> system('ls')

That shouldn't happen, unless you use e.g. `ls --color=always`, which is generally not recommended. So are you sure this still happens? Please provide a reproducible example of such.

**Comment by Shane-XB-Qian** [comment]
i do not think so (xx 'generally' xx), besides that's just a example, or not ls, but 'echo foo', same.

**Comment by Shane-XB-Qian** [comment]
that was not the seq of color, perhaps you mis-understood .. or whatever maybe me ..

**Comment by chrisbra** [comment]
Yeah perhaps. That's why it would be good to have a reproducible example. I just tried `:echo system('echo "abcd"')` in gui and non-gui mode (xterm and non-xterm). No such sequences.

v8.2.931

So please more specific about your environment and when exactly you see the problem.

**Comment by Shane-XB-Qian** [comment]
![image](https://user-images.githubusercontent.com/13550529/84031818-efcd8b80-a9c8-11ea-8ab6-4516ebbcdca5.png)

**Comment by Shane-XB-Qian** [comment]
that's a little 'terminal' specific, so it's normal if you can not reproduce it ..
// i mean so far no clear list for sure which terminal not but which were ok ..
bram knew the background, or you may try xfce-term if you can find it, or some others terminal ..

**Comment by chrisbra** [comment]
> that's a little 'terminal' specific, so it's normal if you can not reproduce it ..

Well, that's why I asked about your environment. 
I checked my xfce4 terminal: and still don't see it. xfce4-terminal 0.8.92. What is your vim version? Have you tried adjusting your $TERM?

**Comment by Shane-XB-Qian** [comment]
vim --clean, and just that 'echo' or 'let xx = ...' with a 'cr'
lastest vim version, xfce-term was said in earlier comment.
that's not about $TERM set, but was a real xterm or not .. looks.

and looked into src code, looks once mode changed, such seq would output, which to fix such tkt.
looks so, .. but now we need to find a way to overcome such output even it was not a real xterm.

**Comment by chrisbra** [comment]
I find it hard to follow what you say when you always a bit brief. Sorry, please try to be a bit more verbose, document exactly what you did, what you expect and what you see. And mention version numbers (or commit hashes, since Vim changes so fast). 

> that's not about $TERM set, but was a real xterm or not .. looks.

Not sure what you are trying to say here. I thought it happens in xfce4-terminal?

> and looked into src code, looks once mode changed, such seq would output, which to fix such tkt.
looks so, .. 

Sorry, I have no idea what you are trying to say.

**Comment by Shane-XB-Qian** [comment]
> That basic problem is still that t_TI and t_TE are set with sequences that the terminal doesn't understand. We can add some workarounds, but we can't fix that problem. It most likely is caused by setting $TERM to a value starting with "xterm" while the terminal is not actually an xterm.

pls read this comment from bram, or the whole comment history, specially those comments after i joined in this tkt.

VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jun  8 2020 13:33:08)
Included patches: 1-931
Compiled by shane@shanesmbxpro

**Comment by chrisbra** [comment]
> pls read this comment from brame, or the whole comment history, specially those comments after i joined in this tkt.

I have read it, but I still find it hard to understand your words. Please write out complete sentences and do not abbreviate too much. Also, have you tried the workaround of not setting $TERM to `xterm` when you are actually not running a xterm? I would even say, this is not even a workaround, but a configuration issue.

Please also note, that his name is Bram, not brame (which you have misspelled before).

**Comment by Shane-XB-Qian** [comment]
no, no, it's not about $TERM you set, but some terminal emulator e.g xfce-term acted like xterm but it was not a real xterm, such seq looks it's not recognized by it.
sorry i was not born with english, :) but whatever, i just tried to let you or bram know that issue still there, but as i said it can be workaround, so it's ok if/just let it be ..
